src/gt4py/next/ffront/decorator.py

```diff
@@ -244,11 +258,59 @@ def __call__(self, *args: Any, offset_provider: common.OffsetProvider, **kwargs:
                 ctx.run(self.definition_stage.definition, *args, **kwargs)
             return
 
-        self.backend(
-            self.definition_stage,
-            *args,
-            **(kwargs | {"offset_provider": offset_provider}),
+        if self._compiled_program is not None:
+            # TODO(havogt): make offset_provider_type part of the compiled program hash once we have variants
+            # TODO(havogt): measure overhead of canonicalize_arguments
+            program_type = self.past_stage.past_node.type
+            assert isinstance(program_type, ts_ffront.ProgramType)
+            args, kwargs = type_info.canonicalize_arguments(program_type, args, kwargs)
+            self._compiled_program.result()(*args, **kwargs, offset_provider=offset_provider)
```

Comment:
After the first call to `result()` we could save the value directly instead of calling `result()` multiple times. I don't know if it is really worth it just to save one (probably cheap) function call, but if it is, one approach could be something like:
```suggestion
            try:
                self._compiled_program(*args, **kwargs, offset_provider=offset_provider)
            except NotCallableSomethiongError:
                self._compiled_program = self._compiled_program.result()
                self._compiled_program(*args, **kwargs, offset_provider=offset_provider)
```

Reply:
Should we do it? In the profile it wasn't visible.

Reply:
I did it.